src/api/cubrid_log.c

```diff
@@ -0,0 +1,1317 @@
+
+  if (reply_code != NO_ERROR)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_CONNECT);
+    }
+
+  return CUBRID_LOG_SUCCESS;
+
+cubrid_log_error:
+
+  if (recv_data != NULL && recv_data != reply)
+    {
+      free_and_init (recv_data);
+    }
+
+  queue_entry = css_find_queue_entry (g_conn_entry->buffer_queue, rid);
+  if (queue_entry != NULL)
+    {
+      queue_entry->buffer = NULL;
+      css_queue_remove_header_entry_ptr (&g_conn_entry->buffer_queue, queue_entry);
+    }
+
+  return err_code;
+}
+
+/*
+ * cubrid_log_connect_server () -
+ *   return:
+ *   host(in):
+ *   port(in):
+ *   dbname(in):
+ */
+int
+cubrid_log_connect_server (char *host, int port, char *dbname)
+{
+  int err_code;
+
+  if (g_stage != CUBRID_LOG_STAGE_CONFIGURATION)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_FUNC_CALL_STAGE);
+    }
+
+  if (host == NULL)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_HOST);
+    }
+
+  if (port < 0 || port > USHRT_MAX)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_PORT);
+    }
+
+  if (dbname == NULL)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_INVALID_DBNAME);
+    }
+
+  if (cubrid_log_connect_server_internal (host, port, dbname) != CUBRID_LOG_SUCCESS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_CONNECT);
+    }
+
+  if (cubrid_log_send_configurations () != CUBRID_LOG_SUCCESS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_FAILED_CONNECT);
+    }
+
+  g_stage = CUBRID_LOG_STAGE_PREPARATION;
+
+  return CUBRID_LOG_SUCCESS;
+
+cubrid_log_error:
+
+  return err_code;
+}
+
+static int
+cubrid_log_find_start_lsa (time_t timestamp, LOG_LSA * lsa)
+{
+  unsigned short rid = 0;
+
+  OR_ALIGNED_BUF (OR_BIGINT_SIZE) a_request;
+  OR_ALIGNED_BUF (OR_INT_SIZE + OR_LOG_LSA_ALIGNED_SIZE) a_reply;
+  char *request = OR_ALIGNED_BUF_START (a_request), *ptr;
+  char *reply = OR_ALIGNED_BUF_START (a_reply);
+  int request_size = OR_ALIGNED_BUF_SIZE (a_request);
+  int reply_size = OR_ALIGNED_BUF_SIZE (a_reply), reply_code;
+
+  char *recv_data = NULL;
+  int recv_data_size;
+
+  CSS_QUEUE_ENTRY *queue_entry;
+  int err_code;
+
+  or_pack_int64 (request, (INT64) timestamp);
+
+  if (css_send_request_with_data_buffer
+      (g_conn_entry, NET_SERVER_LOG_READER_GET_LSA, &rid, request, request_size, reply, reply_size) != NO_ERRORS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_LSA_NOT_FOUND);
+    }
+
+  if (css_receive_data (g_conn_entry, rid, &recv_data, &recv_data_size, g_extraction_timeout * 1000) != NO_ERRORS)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_LSA_NOT_FOUND);
+    }
+
+  if (recv_data == NULL || recv_data_size != OR_INT_SIZE + OR_LOG_LSA_ALIGNED_SIZE)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_LSA_NOT_FOUND);
+    }
+
+  ptr = or_unpack_int (recv_data, &reply_code);
+
+  if (reply_code != NO_ERROR)
+    {
+      CUBRID_LOG_ERROR_HANDLING (CUBRID_LOG_LSA_NOT_FOUND);
+    }
+
+  or_unpack_log_lsa (ptr, lsa);
+
```

Comment:
In server side, lsa will be converted to uint64_t and packed with or_pack_int64()